test/test_main.py

```diff
@@ -138,6 +139,64 @@ def __init__(self, framework):
         self.assertFalse(warn_cm)
 
 
+@patch('sys.argv', new=("hooks/config-changed",))
+class TestDispatch(unittest.TestCase):
+    def _check(self, *, with_dispatch=False, dispatch_path=''):
+        """Helper for below tests."""
+        class MyCharm(CharmBase):
+            def __init__(self, framework):
+                super().__init__(framework)
+
+        fake_environ = {
+            'JUJU_UNIT_NAME': 'test_main/0',
+            'JUJU_MODEL_NAME': 'mymodel',
+        }
+        if dispatch_path:
+            fake_environ['JUJU_DISPATCH_PATH'] = dispatch_path
+            fake_environ['JUJU_VERSION'] = '2.8.0'
+        else:
+            fake_environ['JUJU_VERSION'] = '2.7.0'
+        fake_script(self, 'juju-log', "exit 0")
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            tmpdir = Path(tmpdir)
+            fake_metadata = tmpdir / 'metadata.yaml'
+            with fake_metadata.open('wb') as fh:
+                fh.write(b'name: test')
+            if with_dispatch:
+                dispatch = tmpdir / 'dispatch'
+                with dispatch.open('wt', encoding='utf8') as fh:
+                    fh.write('')
+                    os.fchmod(fh.fileno(), 0o755)
+
+            with patch.dict(os.environ, fake_environ):
+                with patch('ops.main._emit_charm_event') as mock_charm_event:
+                    with patch('ops.main._get_charm_dir') as mock_charmdir:
+                        mock_charmdir.return_value = tmpdir
+                        main(MyCharm)
+
+        self.assertEqual(mock_charm_event.call_count, 1)
+        return mock_charm_event.call_args[0][1]
+
+    def test_most_legacy(self):
+        """Without dispatch, sys.argv[0] is used
+        """
```

Comment:
Please (here and below) end the docstring with a dot and the triple-quote in the same line.